varname/ignore.py

```diff
@@ -176,42 +172,56 @@ def __init__(self, ignore: Tuple[ModuleType, str]) -> None:
         # check uniqueness of qualname
         self.qname_checked = False
         modfile = getattr(self.ignore[0], '__file__', None)
-        if modfile and path.isfile(modfile):
+        if modfile:
             source = Source.for_filename(modfile, self.ignore[0].__dict__)
-            nobj = list(source._qualnames.values()).count(self.ignore[1])
-            # don't check for non-existence, as it could contain wildcards
-            assert nobj <= 1, (
+            self._check_qualname_by_source(source)
+            self.qname_checked = True
+
+    @staticmethod
+    @lru_cache()
+    def _source_from_frame(frame: FrameType) -> Source:
+        """Get source object from the given frame"""
+        return Source.for_frame(frame)
+
+    def _check_qualname_by_source(self, source: Source) -> None:
+        """Check if qualname references to multiple objects"""
+        if not source.tree:
+            # there is no way to check
+            return
+        nobj = list(source._qualnames.values()).count(self.ignore[1])
+        # don't check for non-existence, as it could contain wildcards
+        if nobj > 1:
+            raise QualnameNonUniqueError(
                 f"Qualname {self.ignore[1]!r} in "
                 f"{self.ignore[0].__name__!r} refers to multiple objects."
             )
-            self.qname_checked = True
+
+    @lru_cache()
+    def _check_qualname_by_frame(self, frame: FrameType) -> None:
```

Comment:
Sorry, this isn't what I meant. The arguments to lru_cache should be long lived objects that will be encountered many times for the cache to work well, and neither `self` nor `frame` fits that. Similarly for `_source_from_frame`, but also `Source` has its own caching and doesn't really need the extra layer.

Make a cached free function `_check_qualname_by_source(source, module, qualname)` and then call `_check_qualname_by_source(source, *self.ignore)` in both places. Then there is no need for `qname_checked` and the cache will work across many invocations of `varname()`.

Reply:
> Make a cached free function `_check_qualname_by_source(source, module, qualname)` and then call `_check_qualname_by_source(source, *self.ignore)` in both places. 

Is second `_check_qualname_by_source` a typo?